server/queue.lua

```diff
@@ -0,0 +1,239 @@
+---@class QueueType
+---@field name string
+---@field color string?
+---@field predicate (fun(source: Source): boolean)? 
+
+if GetConvar('qbx:enablequeue', 'true') == 'false' then return end
+
+-- Disable hardcap because it kicks the player when the server is full
+
+---@param resource string
+AddEventHandler('onResourceStarting', function(resource)
+    if resource == 'hardcap' then
+        lib.print.info('Preventing hardcap from starting...')
+        CancelEvent()
+    end
+end)
+
+if GetResourceState('hardcap'):find('start') then
+    lib.print.info('Stopping hardcap...')
+    StopResource('hardcap')
+end
+
+-- Queue code
+
+local config = require 'config.server'.queue
+local maxPlayers = GlobalState.MaxPlayers
+
+---@class SubQueue
+---@field size integer
+---@field positions table<string, integer>
+---@field enqueue fun(self: SubQueue, license: string)
+---@field dequeue fun(self: SubQueue, license: string)
+
+local SubQueue = {}
+local prototype = {}
+local totalSize = 0
+
+---@return SubQueue
+function SubQueue.new()
+    local self = setmetatable({}, { __index = prototype })
+    self.size = 0
+    self.positions = {}
+    return self
+end
+
+---@param self SubQueue
+---@param license string
+function prototype.enqueue(self, license)
+    self.size += 1
+    totalSize += 1
+    self.positions[license] = self.size
```

Comment:
The tradeoff of having license to position is that a less performant loop has to be done when a player is removed from the queue, but the tradeoff of having only position to license is that each time we want to know the position of a player we need to loop. Meaning either we make a less performant loop (since it's done through lua) every time a player disconnects or a more performant loop (I've imagined that `table.remove` would be used since iirc it uses a C loop which theoretically is faster) every second for every player.

I've mapped it license to position because the way I've imagined it is that each player knows their position in the queue and nothing more. I think the only good reason to store position to license/source would be if we need to reveal to a player or an external API what the player in a certain position is, but at the moment the queue does neither, so no reason to store that information currently.

Since the player is updated on their status every second, the position of the player in the queue has to be known every second, but currently it is never asked in what position in line another player is.